File: src/harpy/ast/statements/return_statement.py
from typing import override

from ..expressions import Expression
from .statement import Statement


class ReturnStatement(Statement):
    """Represents a `return` statement with optional return value."""

    _retval: Expression | None

    def __init__(self, retval: Expression | None):
        self._retval = retval

    @override
    def print(self) -> str:
        return "return" + (
            " " + self._retval.print() if self._retval is not None else ""
        )

    def return_value(self) -> Expression | None:
        return self._retval

File: Harpy/AST/Statements/ReturnStatement.cs
using Harpy.AST.Expressions;

namespace Harpy.AST.Statements;

public class ReturnStatement : Statement
{
    private readonly Expression? _returnValue;

    public ReturnStatement(Expression? returnValue) : base([])
    {
        _returnValue = returnValue;

        if (_returnValue == null) return;
        _returnValue.Parent = this;
        Children.Add(_returnValue);
    }

    public override string PrettyPrint(int indent = 0)
    {
        var result = NodeLine(indent) + "ReturnStatement(";
        
        if (_returnValue != null)
        {
            result += "\n" + BlankLine(indent + 1) + "returnValue\n" + ChildNodeLine(indent + 1) + 
                      _returnValue.PrettyPrint(indent + 2) + "\n" + BlankLine(indent);
        }
        
        return result + ")";
    }

    public Expression? ReturnValue()
    {
        return _returnValue;
    }
}
